riptide-stream/src/main/java/org/zalando/riptide/stream/StreamConverter.java

```diff
@@ -106,49 +105,52 @@ public boolean canWrite(Type type, Class<?> clazz, MediaType mediaType) {
     }
 
     @Override
-    public T read(Class<? extends T> clazz, HttpInputMessage inputMessage)
+    public T read(final Class<? extends T> clazz, final HttpInputMessage inputMessage)
             throws IOException, HttpMessageNotReadableException {
-        JavaType javaType = getJavaType(clazz, null);
+        final JavaType javaType = getJavaType(clazz, null);
         return read(javaType, inputMessage);
     }
 
     @Override
-    public T read(Type type, Class<?> contextClass, HttpInputMessage inputMessage)
+    public T read(final Type type, final Class<?> contextClass, final HttpInputMessage inputMessage)
             throws IOException, HttpMessageNotReadableException {
-        JavaType javaType = getJavaType(type, contextClass);
+        final JavaType javaType = getJavaType(type, contextClass);
         return read(javaType, inputMessage);
     }
 
-    @SuppressWarnings("unchecked")
-    private T read(JavaType javaType, HttpInputMessage inputMessage) {
+    private T read(final JavaType javaType, final HttpInputMessage inputMessage) {
         try {
             if (Stream.class.isAssignableFrom(javaType.getRawClass())) {
-                return (T) StreamSupport.stream(split(javaType, inputMessage), false);
+                return stream(javaType.containedType(0), input(inputMessage));
             }
             return mapper.readValue(inputMessage.getBody(), javaType);
         } catch (IOException ex) {
             throw new HttpMessageNotReadableException("Could not read document: " + ex.getMessage(), ex);
         }
     }
 
-    private StreamSpliterator<Object> split(JavaType javaType, HttpInputMessage inputMessage)
+    @SuppressWarnings("unchecked")
+    private T stream(final JavaType javaType, final InputStream stream)
             throws IOException, JsonParseException {
+        final JsonParser parser = mapper.getFactory().createParser(stream);
+        final StreamSpliterator<Object> split = new StreamSpliterator<>(javaType, parser, mapper);
+        return (T) StreamSupport.stream(split, false).onClose(split::close);
```

Comment:
Why adding a close method to the Spliterator? We're creating the parser here, so I'd expect:

```java
return (T) StreamSupport.stream(split, false).onClose(parser::close);
```